File: benchmarks/micro/PureCallGlobalForwarding.c
#include <stdint.h>
#include <unistd.h>

static int64_t counter = 1;

__attribute__((noinline))
static int64_t mix(int64_t value) {
    return (value ^ (value >> 7)) + 17;
}

static int64_t accumulate(int64_t salt) {
    counter = (counter + salt) % 1000000007;
    int64_t first = counter;
    int64_t mixed = mix(first);
    int64_t second = counter;
    return mixed + first + second;
}

int main(void) {
    counter = ((int64_t)getpid() % 31) + 1;
    int64_t checksum = 0;

    for (int32_t i = 0; i < 200000; i += 1) {
        checksum += accumulate((int64_t)(i % 97));
        checksum %= 1000000007;
    }

    return checksum <= 0 ? 1 : 0;
}

File: benchmarks/micro/PureCallGlobalForwarding.rs
static mut COUNTER: i64 = 1;

#[inline(never)]
fn mix(value: i64) -> i64 {
    (value ^ (value >> 7)) + 17
}

fn accumulate(salt: i64) -> i64 {
    unsafe {
        COUNTER = (COUNTER + salt) % 1_000_000_007;
        let first = COUNTER;
        let mixed = mix(first);
        let second = COUNTER;
        mixed + first + second
    }
}

fn main() {
    unsafe {
        COUNTER = (std::process::id() as i64 % 31) + 1;
    }

    let mut checksum = 0_i64;

    for i in 0_i32..200_000 {
        checksum += accumulate((i % 97) as i64);
        checksum %= 1_000_000_007;
    }

    if checksum <= 0 {
        std::process::exit(1);
    }
}
